pkg/applicationserver/linking.go

```diff
@@ -34,34 +34,41 @@ import (
 )
 
 func (as *ApplicationServer) linkAll(ctx context.Context) error {
-	return as.linkRegistry.Range(
-		ctx,
-		[]string{
-			"network_server_address",
-			"api_key",
-			"default_formatters",
-		},
-		func(ctx context.Context, ids ttnpb.ApplicationIdentifiers, target *ttnpb.ApplicationLink) bool {
-			as.startLinkTask(ctx, ids, target)
+	return as.linkRegistry.Range(ctx, nil,
+		func(ctx context.Context, ids ttnpb.ApplicationIdentifiers, _ *ttnpb.ApplicationLink) bool {
+			as.startLinkTask(ctx, ids)
 			return true
 		},
 	)
 }
 
 var linkBackoff = []time.Duration{100 * time.Millisecond, 1 * time.Second, 10 * time.Second}
 
-func (as *ApplicationServer) startLinkTask(ctx context.Context, ids ttnpb.ApplicationIdentifiers, target *ttnpb.ApplicationLink) {
+func (as *ApplicationServer) startLinkTask(ctx context.Context, ids ttnpb.ApplicationIdentifiers) {
 	ctx = log.NewContextWithField(ctx, "application_uid", unique.ID(ctx, ids))
 	as.StartTask(ctx, "link", func(ctx context.Context) error {
-		err := as.link(ctx, ids, target)
+		target, err := as.linkRegistry.Get(ctx, ids, []string{
+			"network_server_address",
+			"api_key",
+			"default_formatters",
+		})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				log.FromContext(ctx).WithError(err).Error("Failed to get link")
```

Comment:
At this point, something really bad is going on, i.e. a registry failure. We may want to retry this.

If the link is not found, we return gracefully to stop the task, i.e. the link has been deleted.